### PR TITLE
make gpus=str in Trainer consistent with command line parsing of string

### LINKED ISSUE #6228: cli: Confused on (str, int, List[int]) variants for argparse for --gpus flag?
## 🐛 Bug

A colleague (@siyuanfeng-tri) and I sometimes get confused on how the `--gpus` flag is to be interpreted by `argparse`. I see the following docs:
https://pytorch-lightning.readthedocs.io/en/1.2.1/advanced/multi_gpu.html#select-gpu-devices

But we're sometimes confused about when `argparse` interpretation will either assume it's the count of the gpu (int/str) or the device index (List[int]).

Are there docs for this? If not, can that be clarified somehow?

## Please reproduce using the BoringModel

see notebook

The main complaint is that `gpus=3` implies `gpus=[3]`, while `gpus="3"` implies `gpus=[0,1,2]`.
Mix that with implicit conversion from `argparse` from `str` to `int`, and you get a kinda weird public interface.

### To Reproduce

example notebook: https://colab.research.google.com/drive/1pe9_F2S73-gQ3hOeh_MMiGhmbXmGURDQ?usp=sharing

### Expected behavior

Less confusing / more explicit options? (maybe my complaint is with weird implicit behavior of `Trainer(gpus=...)`?)

### Environment

 - PyTorch Version (e.g., 1.7.1):
 - OS: Ubuntu 18.04
 - How you installed PyTorch: `pip`
 - Python version: 3.6.9
 - CUDA/cuDNN version: N/A
 - GPU models and configuration: N/A
 - Any other relevant information: N/A

### Additional context

N/A

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #6228
Closes #7236

Current behavior:

```python
# Selects GPU device 3
Trainer(gpus="3")
```

Desired behavior:

```python
# Selects first 3 GPUS
Trainer(gpus="3")
```

This change makes the string type parsing for the gpus flag consistent with how the command line arguments are parsed.
**This PR announces the change today (from 1.3), the changes will take effect in 1.5 (breaking change). **

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/models/test_gpu.py [hunk 1]
```diff
@@ -171,8 +171,10 @@ def test_determine_root_gpu_device(gpus, expected_root_gpu):
...
     pytest.param([1, 3], [1, 3]),
     pytest.param((1, 3), [1, 3]),
-    pytest.param('0', [0]),
-    pytest.param('3', [3]),
+    pytest.param('0', [0]),  # TODO: in v1.5 replace this by the commented line below
+    # pytest.param('0', None),
+    pytest.param('3', [3]),  # TODO: in v1.5 replace this by the commented line below
+    # pytest.param('3', [0, 1, 2]),
     pytest.param('1, 3', [1, 3]),
     pytest.param('2,', [2]),
```
Comment (carmocca) [inline tests/models/test_gpu.py:179]: Can you add a test for `'[2]'`?

**Review by carmocca — COMMENTED** [review]
No strong feelings for either a) or b). But if you choose b), you could leave all this logic for the future behaviour implemented already by checking the current lightning version. Just in case we forget about this.

Then when 1.5 comes we just have to remove the old code instead of resolving TODOs

File: tests/models/test_gpu.py (hunk 1 above)
Comment (awaelchli) [reply tests/models/test_gpu.py:179]: this is not implemented. we don't document it anywhere, do we?
is it something we should support?

Comment (carmocca) [reply tests/models/test_gpu.py:179]: nvm, got misled by https://github.com/PyTorchLightning/pytorch-lightning/issues/6228#issuecomment-787854844

**Comment by edenlightning** [comment]
I really think we should avoid breaking changes, esp when this is unexpected.

**Comment by awaelchli** [comment]
@edenlightning so you want me to close this PR? this would mean we keep the current behavior, nothing changes, and also close the linked issue.

**Comment by Borda** [comment]
@awaelchli still draft?

**Comment by awaelchli** [comment]
@Borda it's 95% done, just waiting for a go/no-go decision here. 
If backward incompatible changes are not acceptable, this PR can be closed. @edenlightning

File: pytorch_lightning/utilities/device_parser.py [hunk 2]
```diff
@@ -107,13 +112,25 @@ def parse_tpu_cores(tpu_cores: Union[int, str, List]) -> Optional[Union[List[int
...
+    if not isinstance(s, str):
         return s
+    if s == '-1':
+        return -1
+    elif ',' in s:
+        return [int(x.strip()) for x in s.split(',') if len(x) > 0]
+    else:
+        num_gpus = int(s.strip())
+        if _compare_version("pytorch_lightning", operator.lt, "1.5"):
+            # TODO: remove warning in v1.5 and update docs in docs/advanced/multi-gpu.rst regarding GPU selection
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/device_parser.py:124]: I don't think it's required here if we are mentioning it in `test_remove_1-5.py`

File: pytorch_lightning/utilities/device_parser.py [hunk 3]
```diff
@@ -107,13 +112,25 @@ def parse_tpu_cores(tpu_cores: Union[int, str, List]) -> Optional[Union[List[int
...
+            rank_zero_warn(
+                f"Parsing of the Trainer argument gpus='{s}' (string) will change in the future."
+                f" In the current version of Lightning, this will select"
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/device_parser.py:127]:
```suggestion
                " In the current version of Lightning, this will select"
```

**Review by kaushikb11 — APPROVED** [review]
LGTM!

File: pytorch_lightning/utilities/device_parser.py [hunk 4]
```diff
@@ -107,13 +112,25 @@ def parse_tpu_cores(tpu_cores: Union[int, str, List]) -> Optional[Union[List[int
...
+                f" CUDA device with index {num_gpus}, but from v1.5 it will select gpus"
+                f" {list(range(num_gpus))}.",
+                DeprecationWarning,
+            )
+            return [num_gpus]
+        return num_gpus
```
Comment (kaushikb11) [inline pytorch_lightning/utilities/device_parser.py:133]: @awaelchli I wonder what values reach this point.

Comment (awaelchli) [reply pytorch_lightning/utilities/device_parser.py:133]: The last return statement? Currently, this cannot be reached, because we are below version 1.5.
But from version 1.5 it will hit the last return, i.e., `Trainer(gpus="3")` will be interpreted as the number of gpus, just like `Trainer(gpus=int)` does.

File: pytorch_lightning/utilities/device_parser.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/device_parser.py:124]: okay

File: pytorch_lightning/utilities/device_parser.py (hunk 4 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/device_parser.py:132]: I did it this way so that in 1.5 all we have to do is delete the if block above and everything else will be correct.

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/utilities/device_parser.py (hunk 4 above)
Comment (kaushikb11) [reply pytorch_lightning/utilities/device_parser.py:132]: Sounds good, thanks!

File: pytorch_lightning/utilities/device_parser.py [hunk 5]
```diff
@@ -107,13 +112,24 @@ def parse_tpu_cores(tpu_cores: Union[int, str, List]) -> Optional[Union[List[int
...
+    if s == '-1':
+        return -1
+    elif ',' in s:
+        return [int(x.strip()) for x in s.split(',') if len(x) > 0]
+    else:
+        num_gpus = int(s.strip())
+        if _compare_version("pytorch_lightning", operator.lt, "1.5"):
+            rank_zero_warn(
+                f"Parsing of the Trainer argument gpus='{s}' (string) will change in the future."
+                " In the current version of Lightning, this will select"
```
Comment (EricCousineau-TRI) [inline pytorch_lightning/utilities/device_parser.py:126]: BTW Consider explicitly mentioning `int`; e.g. "but from v1.5 it will it will select gpus (same as gpus=d (int)) [...]"

**Comment by carmocca** [comment]
@awaelchli I see you modified the milestone here. Note that if this goes into 1.4 you'll have to update the deprecation warnings

**Comment by awaelchli** [comment]
Yes this got pushed back so I'll pick it up after 1.3. Thanks.

File: docs/source/advanced/multi_gpu.rst [hunk 6]
```diff
@@ -226,13 +226,17 @@ Note in particular the difference between `gpus=0`, `gpus=[0]` and `gpus="0"`.
...
+| "3"           | str       | [3]                 | GPU 3 (will change in v1.5)     |
 +---------------+-----------+---------------------+---------------------------------+
 | "1, 3"        | str       | [1, 3]              | GPUs 1 and 3                    |
 +---------------+-----------+---------------------+---------------------------------+
 | "-1"          | str       | [0, 1, 2, ...]      | all available GPUs              |
 +---------------+-----------+---------------------+---------------------------------+
 
+.. warning::
+    The behavior for :code:`gpus="3"` (str) will change. Currently it selects the GPU with index 3, but will
+    select the first 3 GPUs from v1.5.
```
Comment (Borda) [inline docs/source/advanced/multi_gpu.rst:238]: why are we announcing a change in 2 versions ahead?

**Review by Borda — APPROVED** [review]

File: docs/source/advanced/multi_gpu.rst (hunk 6 above)
Comment (awaelchli) [reply docs/source/advanced/multi_gpu.rst:238]: are you saying one version is enough? I followed deprecation guideline of two versions.

File: tests/models/test_gpu.py [hunk 7]
```diff
@@ -22,12 +23,14 @@
...
 from pytorch_lightning.utilities import device_parser
 from pytorch_lightning.utilities.exceptions import MisconfigurationException
+from pytorch_lightning.utilities.imports import _compare_version
 from tests.helpers import BoringModel
 from tests.helpers.datamodules import ClassifDataModule
 from tests.helpers.imports import Batch, Dataset, Example, Field, LabelField
 from tests.helpers.runif import RunIf
 from tests.helpers.simple_models import ClassificationModel
 
+PL_VERSION_LT_1_5 = _compare_version("pytorch_lightning", operator.lt, "1.5")
```
Comment (tchaton) [inline tests/models/test_gpu.py:33]: Nice !

**Review by tchaton — APPROVED** [review]
I am for it. BC should be prevented when possible, unless it is to correct a behaviour.

File: pytorch_lightning/utilities/device_parser.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/device_parser.py:126]: done, thanks!
